### PR TITLE
stuckness fixes

### PR DESCRIPTION
**Pull request by zneedell**
Two attempts to fix stuckness in production sfbay runs. The combination of these two allowed the runs to successfully get through many iterations.

- Push back departure times until `tick + schedulerParallelismWindow` when the agent begins an activity so delayed that it may interfere with their next scheduled trip
- Only allow charging events of duration `>= schedulerParallelismWindow` 

Presumably both of these could be more thoroughly addressed through changes to the protocols, but I think that will get really messy and these fixes don't have any meaningful impact on behavior/outcomes as long as the parallelism window isn't huge.

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/3510)

### CODE REVIEW

**Review by nikolayilyin — APPROVED** [review]
lgtm

**Comment by JustinPihony** [comment]
When run against `production/sfbay/smart/smart-baseline.conf` from `git submodule update --init --remote production/sfbay` - https://beam-ci.tk/job/application-sfbay/462/console :

```
02:09:18 Caused by: java.util.NoSuchElementException: head of empty list
02:09:18 	at scala.collection.immutable.Nil$.head(List.scala:469)
02:09:18 	at scala.collection.immutable.Nil$.head(List.scala:466)
02:09:18 	at beam.agentsim.infrastructure.InfrastructureUtils$.buildParkingAndChargingNetworks(InfrastructureUtils.scala:105)
02:09:18 	at beam.sim.BeamMobsimIteration.<init>(BeamMobsim.scala:397)
02:09:18 	at beam.sim.BeamMobsim.$anonfun$run$5(BeamMobsim.scala:167)
02:09:18 	at akka.actor.TypedCreatorFunctionConsumer.produce(IndirectActorProducer.scala:91)
02:09:18 	at akka.actor.Props.newActor(Props.scala:226)
02:09:18 	at akka.actor.ActorCell.newActor(ActorCell.scala:613)
02:09:18 	at akka.actor.ActorCell.create(ActorCell.scala:640)
02:09:18 	... 9 common frames omitted
```

File: src/main/scala/beam/agentsim/agents/PersonAgent.scala [hunk 1]
```diff
@@ -1393,9 +1406,22 @@ class PersonAgent(
           )
           eventsManager.processEvent(activityStartEvent)
 
+          val nextLegDepartureTime =
+            if (activityEndTime >= tick + beamServices.beamConfig.beam.agentsim.schedulerParallelismWindow) {
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/PersonAgent.scala]: Why is this one `>=` when the others are only `>`?

**Comment by JustinPihony** [comment]
Also, the main develop branch fails `beam.sim.population.PopulationAdjustmentSpec.PopulationAdjustment should logs excluded modes and alarms not all persons have required attribute`

```
ArrayBuffer() was not equal to Array((INFO,Modes excluded:), (INFO,car -> 3), (INFO,bike -> 2), (ERROR,Not all agents have person attributes - is attributes file missing ?))
      org.scalatest.exceptions.TestFailedException: ArrayBuffer() was not equal to Array((INFO,Modes excluded:), (INFO,car -> 3), (INFO,bike -> 2), (ERROR,Not all agents have person attributes - is attributes file missing ?))
      at org.scalatest.matchers.MatchersHelper$.indicateFailure(MatchersHelper.scala:402)
      at org.scalatest.matchers.should.Matchers$AnyShouldWrapper.shouldBe(Matchers.scala:7021)
      at beam.sim.population.PopulationAdjustmentSpec.verifyLogging(PopulationAdjustmentSpec.scala:120)
      at beam.sim.population.PopulationAdjustmentSpec.$anonfun$new$8(PopulationAdjustmentSpec.scala:102)
      at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)
      at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
      at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
      at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
      at org.scalatest.Transformer.apply(Transformer.scala:22)
      at org.scalatest.Transformer.apply(Transformer.scala:20)
      at org.scalatest.wordspec.AnyWordSpecLike$$anon$3.apply(AnyWordSpecLike.scala:1227)
      at org.scalatest.TestSuite.withFixture(TestSuite.scala:196)
      at org.scalatest.TestSuite.withFixture$(TestSuite.scala:195)
      at org.scalatest.wordspec.AnyWordSpec.withFixture(AnyWordSpec.scala:1879)
      at org.scalatest.wordspec.AnyWordSpecLike.invokeWithFixture$1(AnyWordSpecLike.scala:1225)
      at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTest$1(AnyWordSpecLike.scala:1237)
      at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
      at org.scalatest.wordspec.AnyWordSpecLike.runTest(AnyWordSpecLike.scala:1237)
      at org.scalatest.wordspec.AnyWordSpecLike.runTest$(AnyWordSpecLike.scala:1219)
      at beam.sim.population.PopulationAdjustmentSpec.org$scalatest$BeforeAndAfterEach$$super$runTest(PopulationAdjustmentSpec.scala:21)
      at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
      at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
      at beam.sim.population.PopulationAdjustmentSpec.runTest(PopulationAdjustmentSpec.scala:21)
      at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$runTests$1(AnyWordSpecLike.scala:1296)
      at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:413)
      at scala.collection.immutable.List.foreach(List.scala:431)
      at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
      at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:390)
      at org.scalatest.SuperEngine.$anonfun$runTestsInBranch$1(Engine.scala:427)
      at scala.collection.immutable.List.foreach(List.scala:431)
      at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
      at org.scalatest.SuperEngine.runTestsInBranch(Engine.scala:396)
      at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:475)
      at org.scalatest.wordspec.AnyWordSpecLike.runTests(AnyWordSpecLike.scala:1296)
      at org.scalatest.wordspec.AnyWordSpecLike.runTests$(AnyWordSpecLike.scala:1295)
      at org.scalatest.wordspec.AnyWordSpec.runTests(AnyWordSpec.scala:1879)
      at org.scalatest.Suite.run(Suite.scala:1112)
      at org.scalatest.Suite.run$(Suite.scala:1094)
      at org.scalatest.wordspec.AnyWordSpec.org$scalatest$wordspec$AnyWordSpecLike$$super$run(AnyWordSpec.scala:1879)
      at org.scalatest.wordspec.AnyWordSpecLike.$anonfun$run$1(AnyWordSpecLike.scala:1341)
      at org.scalatest.SuperEngine.runImpl(Engine.scala:535)
      at org.scalatest.wordspec.AnyWordSpecLike.run(AnyWordSpecLike.scala:1341)
      at org.scalatest.wordspec.AnyWordSpecLike.run$(AnyWordSpecLike.scala:1339)
      at org.scalatest.wordspec.AnyWordSpec.run(AnyWordSpec.scala:1879)
      at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:45)
      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
      at java.util.concurrent.FutureTask.run(FutureTask.java:266)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
      at java.lang.Thread.run(Thread.java:748)
```

**Review by JustinPihony — DISMISSED** [review]
Blocking to double check the equality question

**Comment by Xuan-1998** [comment]
Test!

**Pull request merged**